packages/playwright-crawler/src/internals/adaptive-playwright-crawler.ts

```diff
@@ -0,0 +1,263 @@
+        this.enqueuedUrls.push(...(options?.urls?.map((url) => ({ url, label: options.label, options })) ?? []));
+    };
+
+    addRequests: RestrictedCrawlingContext['addRequests'] = async (requests, options = {}) => {
+        for (const request of requests) {
+            if (typeof request === 'object' && 'requestsFromUrl' in request) {
+                this.requestListUrls.push({ url: request.url, label: request.label, request, options });
+            } else if (typeof request === 'string') {
+                this.requestUrls.push({ url: request, request: { url: request }, options });
+            } else {
+                this.requestUrls.push({ url: request.url, label: request.label, request, options });
+            }
+        }
+    };
+
+    useState: RestrictedCrawlingContext['useState'] = async (defaultValue) => {
+        // @ts-ignore
+        const key = BasicCrawler.CRAWLEE_STATE_KEY;
+        const store = await this.getKeyValueStore(undefined);
+
+        return await store.getAutoSavedValue(key, defaultValue);
+    };
+
+    private idOrDefault = (idOrName?: string): string => idOrName ?? this.config.get('defaultKeyValueStoreId');
+
+    private getKeyValueStoreChangedValue = (idOrName: string | undefined, key: string) => {
+        const id = this.idOrDefault(idOrName);
+        this.keyValueStoreChanges[id] ??= {};
+        return this.keyValueStoreChanges[id][key]?.changedValue ?? null;
+    };
+
+    private setKeyValueStoreChangedValue = (idOrName: string | undefined, key: string, changedValue: unknown, options?: RecordOptions) => {
+        const id = this.idOrDefault(idOrName);
+        this.keyValueStoreChanges[id] ??= {};
+        this.keyValueStoreChanges[id][key] = { changedValue, options };
+    };
+
+    getKeyValueStore: RestrictedCrawlingContext['getKeyValueStore'] = async (idOrName) => {
+        const store = await KeyValueStore.open(idOrName, { config: this.config });
+
+        return {
+            id: this.idOrDefault(idOrName),
+            name: idOrName,
+            getValue: async (key) => this.getKeyValueStoreChangedValue(idOrName, key) ?? await store.getValue(key),
+            getAutoSavedValue: async <T extends Dictionary = Dictionary>(key: string, defaultValue: T = {} as T) => {
+                let value = this.getKeyValueStoreChangedValue(idOrName, key);
+                if (value === null) {
+                    value = await store.getValue(key) ?? defaultValue;
+                    this.setKeyValueStoreChangedValue(idOrName, key, value);
+                }
+
+                return value as T;
+            },
+            setValue: async (key, value, options) => {
+                this.setKeyValueStoreChangedValue(idOrName, key, value, options);
+            },
+        };
+    };
+}
+
+interface AdaptivePlaywrightCrawlerContext extends RestrictedCrawlingContext {
+    dom: CheerioPortadom<Cheerio<Document>> | PlaywrightLocatorPortadom;
```

Comment:
I'm not sure if portadom is a good idea. Check out it's [npm page](https://www.npmjs.com/package/portadom) for usage examples.

A viable alternative would be an async function that returns a cheerio object, but waits on a locator in playwright context.

I am of course open to any suggestions.

Reply:
Depending on a potentially unstable API is a strong point against it, yes. What do you think about the ergonomics of the portadom API?